pallets/parachain-staking/src/types.rs

```diff
@@ -0,0 +1,1673 @@
+    /// Only call if lowest top delegation is less than delegation.amount || !top_full
+    pub fn add_top_delegation<T: Config>(
+        &mut self,
+        candidate: &T::AccountId,
+        delegation: Bond<T::AccountId, BalanceOf<T>>,
+    ) -> Option<Balance>
+    where
+        BalanceOf<T>: Into<Balance> + From<Balance>,
+    {
+        let mut less_total_staked = None;
+        let mut top_delegations = <TopDelegations<T>>::get(candidate)
+            .expect("CandidateInfo existence => TopDelegations existence");
+        let max_top_delegations_per_candidate = T::MaxTopDelegationsPerCandidate::get();
+        if top_delegations.delegations.len() as u32 == max_top_delegations_per_candidate {
+            // pop lowest top delegation
+            let new_bottom_delegation = top_delegations.delegations.pop().expect("");
+            top_delegations.total = top_delegations
+                .total
+                .saturating_sub(new_bottom_delegation.amount);
+            if matches!(self.bottom_capacity, CapacityStatus::Full) {
+                less_total_staked = Some(self.lowest_bottom_delegation_amount);
+            }
+            self.add_bottom_delegation::<T>(true, candidate, new_bottom_delegation);
+        }
+        // insert into top
+        top_delegations.insert_sorted_greatest_to_least(delegation);
+        // update candidate info
+        self.reset_top_data::<T>(candidate.clone(), &top_delegations);
+        if less_total_staked.is_none() {
+            // only increment delegation count if we are not kicking a bottom delegation
+            self.delegation_count = self.delegation_count.saturating_add(1u32);
+        }
+        <TopDelegations<T>>::insert(&candidate, top_delegations);
+        less_total_staked
+    }
+    /// Add delegation to bottom delegations
+    /// Check before call that if capacity is full, inserted delegation is higher than lowest
+    /// bottom delegation (and if so, need to adjust the total storage item)
+    /// CALLER MUST ensure(lowest_bottom_to_be_kicked.amount < delegation.amount)
+    pub fn add_bottom_delegation<T: Config>(
+        &mut self,
+        bumped_from_top: bool,
+        candidate: &T::AccountId,
+        delegation: Bond<T::AccountId, BalanceOf<T>>,
+    ) where
+        BalanceOf<T>: Into<Balance> + From<Balance>,
+    {
+        let mut bottom_delegations = <BottomDelegations<T>>::get(candidate)
+            .expect("CandidateInfo existence => BottomDelegations existence");
+        // if bottom is full, kick the lowest bottom (which is expected to be lower than input
+        // as per check)
+        let increase_delegation_count = if bottom_delegations.delegations.len() as u32
+            == T::MaxBottomDelegationsPerCandidate::get()
+        {
+            let lowest_bottom_to_be_kicked = bottom_delegations
+                .delegations
+                .pop()
+                .expect("if at full capacity (>0), then >0 bottom delegations exist; qed");
```

Comment:
expect

Reply:
this invariant holds as long as MaxBottomDelegationsPerCandidate is != 0

Which is interesting, because it's not documented. But as it stands - not a bug